File: 345-reverse-vowels-of-a-string/reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        int low = 0;
        int high = s.size()-1;
        set<char> ss = { 'a','e','i','o','u','A','E','I','O','U'};
        while(low<=high)
        {
            while(ss.find(s[low])==ss.end() && low<=high) low++;
            while(ss.find(s[high])==ss.end() && low<=high) high--;
            if(low<=high){
            char c = s[low];
            s[low] = s[high];
            s[high] = c;
            high--;
            low++;
            }
            else break;
        }

        return s;
    }
};

File: 345-reverse-vowels-of-a-string/reverse-vowels-of-a-string.java
class Solution {
    public String reverseVowels(String s) {
        int low = 0;
        int high = s.length() - 1;
        char[] arr = s.toCharArray();

        // set of vowels
        Set<Character> vowels = new HashSet<>(
            Arrays.asList('a','e','i','o','u','A','E','I','O','U')
        );

        while (low < high) {
            // move low forward until we find a vowel
            while (low < high && !vowels.contains(arr[low])) {
                low++;
            }
            // move high backward until we find a vowel
            while (low < high && !vowels.contains(arr[high])) {
                high--;
            }
            // swap vowels
            if (low < high) {
                char temp = arr[low];
                arr[low] = arr[high];
                arr[high] = temp;
                low++;
                high--;
            }
        }
        return new String(arr);
    }
}